librad/src/id/entity.rs

```diff
@@ -0,0 +1,515 @@
+        key: &PublicKey,
+        by: &Signatory,
+        signature: &Signature,
+        resolver: &impl Resolver<User>,
+    ) -> Result<(), Error> {
+        self.check_key(key, by, resolver).await?;
+        if signature.verify(&self.canonical_data()?, key) {
+            Ok(())
+        } else {
+            Err(Error::SignatureVerificationFailed)
+        }
+    }
+
+    pub async fn check_validity(&self, resolver: &impl Resolver<User>) -> Result<(), Error> {
+        let mut keys = HashSet::<PublicKey>::from_iter(self.keys().iter().cloned());
+        let mut users = HashSet::<RadicleUri>::from_iter(self.certifiers().iter().cloned());
+
+        for (k, s) in self.signatures() {
+            self.check_signature(k, &s.by, &s.sig, resolver).await?;
+            match &s.by {
+                Signatory::OwnedKey => {
+                    keys.remove(k);
+                },
+                Signatory::User(user) => {
+                    users.remove(&user);
+                },
+            }
+        }
+        if keys.is_empty() && users.is_empty() {
+            Ok(())
+        } else {
+            Err(Error::SignatureMissing)
+        }
+    }
+
+    pub async fn is_valid(&self, resolver: &impl Resolver<User>) -> bool {
+        self.check_validity(resolver).await.is_ok()
+    }
+
+    pub fn check_update(&self, previous: &Self) -> Result<(), UpdateVerificationError> {
+        if self.revision() <= previous.revision() {
+            return Err(UpdateVerificationError::NonMonotonicRevision);
+        }
+
+        let retained_keys = self.keys().iter().filter(|k| previous.has_key(k)).count();
+        let total_keys = self.keys_count();
+        let added_keys = total_keys - retained_keys;
+        let removed_keys = previous.keys_count() - retained_keys;
+        let quorum_keys = total_keys / 2;
+
+        if added_keys > quorum_keys {
+            return Err(UpdateVerificationError::NoCurrentQuorum);
+        } else if removed_keys > quorum_keys {
+            return Err(UpdateVerificationError::NoPreviousQuorum);
+        }
+
+        let retained_certifiers = self
+            .certifiers()
+            .iter()
+            .filter(|c| previous.has_certifier(c))
+            .count();
+        let total_certifiers = self.certifiers_count();
+        let added_certifiers = total_certifiers - retained_certifiers;
+        let removed_certifiers = previous.certifiers_count() - retained_certifiers;
+        let quorum_certifiers = total_certifiers / 2;
+
+        if added_certifiers > quorum_certifiers {
+            return Err(UpdateVerificationError::NoCurrentQuorum);
+        } else if removed_certifiers > quorum_certifiers {
+            return Err(UpdateVerificationError::NoPreviousQuorum);
+        }
+
+        Ok(())
+    }
+
+    pub async fn check_history<R>(
+        resolver: &impl Resolver<User>,
+        revisions: R,
+    ) -> Result<(), HistoryVerificationError>
+    where
+        R: Stream<Item = Entity<T>> + Unpin,
```

Comment:
Ok, so you're walking the history from the most recent revision, which simplifies some things. However, I'm missing where we'd verify that the hash of the initial version matches the identity hash.

I also think this shouldn't return unit -- we will need to be able to distinguish between (a) history is valid, (b) head requires your signature **or** head quorum is not reached, so reset to head^, (c) history failed verification at revision X, and (d) an error occurred.

Reply:
I added the parent hash verification (sorry, I just forgot it).

The `HistoryVerificationError` type encodes all the different error cases, and if I am not mistaken the only one not explicitly detected is distinguishing the two subcases of (b).

More generally we don't have a "protocol" for handling "draft" entities waiting to be signed.
They need to be moved from device to device (because the private keys needed to sign them should not leave the devices) but it is still not clear if this "movement" should be implemented by committing them to the entity history branch and pulling from other devices or "out of band" outside the git branch replication machinery.

In this code an entity with missing signatures is considered invalid (and the verification error states so).

Reply:
The errors are fine, but we will need to return something in order to implement whatever interaction flow will be used to prompt for a signature. This will also depend on who the verifier is, and we will need to inspect up to which revision (both native and in-doc) we verified. This suggests to return the entity itself, tagged by the quorum state.

Reply:
The errors are fine, but we will need to return something in order to implement whatever interaction flow will be used to prompt for a signature. This will also depend on who the verifier is, and we will need to inspect up to which revision (both native and in-doc) we verified. This suggests to return the entity itself, tagged by the quorum state.

Reply:
I added the "verification status", not verification methods update this status.

Missing signatures are not a "verification error" if you check a single revision: the entity has a status that states that it must be signed but it is otherwise OK.

However missing signatures become errors when checking the full history.